File: LinkedLists/8/leetFlatten.js
/**
 * // Definition for a Node.
 * function Node(val,prev,next,child) {
 *    this.val = val;
 *    this.prev = prev;
 *    this.next = next;
 *    this.child = child;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var flatten = function (head) {
  let cur = head;
  make_links(cur, null)
  return head;
};

var make_links = function (cur, pnext) {
  if (cur) {
    const child = cur.child;
    const next = cur.next;
    if (child) {
      const cnext = next || pnext;
      cur.next = child;
      child.prev = cur;
      cur.child = null;
      make_links(child, cnext);
    }
    make_links(next, pnext);
    if (cur.next === null) {
      cur.next = pnext;
      if (pnext) {
        pnext.prev = cur;
      }
    }
  }
}

File: LinkedLists/8/leetFlatten.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""


class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        cur = head
        self.make_links(cur, None)
        return head

    def make_links(self, cur, pnext):
        if cur:
            child = cur.child
            next = cur.next
            if child:
                cnext = next or pnext
                cur.next = child
                child.prev = cur
                cur.child = None
                self.make_links(child, cnext)
            self.make_links(next, pnext)
            if cur.next == None:
                cur.next = pnext
                if pnext:
                    pnext.prev = cur
